src/doorman/Channel.cpp

```diff
@@ -0,0 +1,393 @@
+                nLeft -= 3;
+                break;
+            case SB: {
+                char sbType;
+                byte *eob, *p;
+                /* Enough room for an SB command? */
+                if (nLeft < 6) {
+                    nLeft = 0;
+                    break;
+                }
+                sbType = *(ptr + 2);
+                //          sbWhat = *(ptr + 3);
+                /* Now read up to the IAC SE */
+                eob = ptr + nLeft - 1;
+                for (p = ptr + 4; p < eob; ++p)
+                    if (*p == IAC && *(p + 1) == SE)
+                        break;
+                /* Found IAC SE? */
+                if (p == eob) {
+                    /* No: skip it all */
+                    nLeft = 0;
+                    break;
+                }
+                *p = '\0';
+                /* Now to decide what to do with this new data */
+                switch (sbType) {
+                case TELOPT_TTYPE:
+                    log_out("[%d] TTYPE = %s", id_, ptr + 4);
+                    if (SupportsAnsi((const char *)ptr + 4))
+                        ansi_ = true;
+                    break;
+                case TELOPT_NAWS:
+                    width_ = ptr[4];
+                    height_ = ptr[6];
+                    log_out("[%d] NAWS = %dx%d", id_, width_, height_);
+                    break;
+                }
+
+                /* Remember eob is 1 byte behind the end of buffer */
+                memmove(ptr, p + 2, nLeft - ((p + 2) - ptr));
+                nLeft -= ((p + 2) - ptr);
+
+                break;
+            }
+
+            default:
+                memmove(ptr, ptr + 2, nLeft - 2);
+                nLeft -= 2;
+                break;
+            }
+
+        } else {
+            ptr++;
+            nRealBytes++;
+            nLeft--;
+        }
+    }
+    /* Now to update the buffer stats */
+    buffer_.resize(nRealBytes);
+
+    /* Second pass - look for whole lines of text */
+    nLeft = buffer_.size();
+    byte *ptr;
+    for (ptr = buffer_.data(); nLeft; ++ptr, --nLeft) {
+        char c;
+        switch (*ptr) {
+        case '\n':
+        case '\r':
+            // TODO this is terribly ghettish. don't need to memmove, and all that. ugh. But write tests first...
+            // then change...
+            c = *ptr; // legacy<-
+            /* Send it to the MUD */
+            mud_->on_client_message(
+                *this, std::string_view(reinterpret_cast<const char *>(buffer_.data()), ptr - buffer_.data()));
+            /* Check for \n\r or \r\n */
+            if (nLeft > 1 && (*(ptr + 1) == '\r' || *(ptr + 1) == '\n') && *(ptr + 1) != c) {
+                memmove(buffer_.data(), ptr + 2, nLeft - 2);
+                nLeft--;
+            } else if (nLeft) {
+                memmove(buffer_.data(), ptr + 1, nLeft - 1);
+            } // else nLeft is zero, so we might as well avoid calling memmove(x,y,
```

Comment:
Not sure I understand this comment. If nLeft is 1, we're going to call memmove(_, _, 0); above anyway; AFAICT what we save with the 'if (nLeft)' check is calling memmove with -1 as the number of bytes. Right?

Reply:
The comment is wrong, and the code is heinous. In fact my IDE warns me that nLeft CANNOT be zero at this point. But it's left exactly as it was before for now... I'm not touching any of the parsing code until I have tests for it all. It hurts too much :)